C/plugins/storage/common/disk_monitor.cpp

```diff
@@ -0,0 +1,159 @@
+{
+	m_logger = Logger::getLogger();
+}
+
+/**
+ * Called periodically to monitor the disk usage
+ *
+ * @param interval	The number of seconds between calls
+ */
+void DiskSpaceMonitor::periodic(int interval)
+{
+	struct statfs stf1, stf2;
+
+	if (!m_started)
+	{
+		if (statfs(m_dbPath1.c_str(), &stf1) != 0)
+		{
+			m_logger->error("Can;t stafs %s", m_dbPath1.c_str());
+			return;
+		}
+		if (statfs(m_dbPath2.c_str(), &stf2) != 0)
+		{
+			m_logger->error("Can;t stafs %s", m_dbPath2.c_str());
+			return;
+		}
+		if (memcmp(&stf1.f_fsid, &stf2.f_fsid, sizeof(fsid_t)) == 0)	// Same filesystem
+		{
+			m_sameDevice = true;
+		}
+		m_started = true;
+	}
+	m_lastCheck += interval;
+
+	if (m_lastCheck < CHECK_THRESHOLD)
+	{
+		// Do not check too frerquenctly
+		return;
+	}
+	m_lastCheck = 0;
+
+	
+	if (statfs(m_dbPath1.c_str(), &stf1) == 0)
+	{
+		unsigned long freespace = (unsigned long)stf1.f_bavail;
+		unsigned long totalspace = (unsigned long)stf1.f_blocks;
+
+		double perc = (double)(freespace  * 100.0) / totalspace;
+
+		if (perc < 5.0)
+		{
+			m_logger->fatal("Disk space is critically low. Urgent action required, continuing may result in data corruption");
+		}
+		else if (perc < 10.0)
+		{
+			m_logger->error("Available disk space is becoming low, please consider releasing more disk space");
+		}
+		if (m_lastPerc1 > 0.0)
+		{
+			double diff = m_lastPerc1 - perc;
+			if (diff > 0.0)
+			{
+				double prediction = (perc * CHECK_THRESHOLD)/ (3600.0 * diff);
+				if (prediction <= 72.0 && m_lastPrediction1 - prediction > 1.0)
+				{
+					m_logger->error("At current rates disk space will be exhausted in %.0f hours", prediction);
+					m_lastPrediction1 = prediction;
+				}
+				else if (prediction / 24.0 <= 14 && (m_lastPrediction1 == 0.0 || m_lastPrediction1 - prediction > 24.0))
+				{
+					m_lastPrediction1 = prediction;
+					m_logger->warn("At current rates disk space will be exhausted in %.1f days", prediction / 24);
+				}
+			}
+			else
+			{
+				m_lastPrediction1 = 0.0;
+			}
+		}
+		m_lastPerc1 = perc;
+
+	}
+	if (m_sameDevice)
+	{
+		return;
+	}
+	if (statfs(m_dbPath2.c_str(), &stf1) == 0)
+	{
+		unsigned long freespace = (unsigned long)stf1.f_bavail;
+		unsigned long totalspace = (unsigned long)stf1.f_blocks;
+
+		double perc = (double)(freespace  * 100.0) / totalspace;
+
+		if (perc < 5.0)
+		{
+			m_logger->fatal("Disk space is critically low. Urgent action required, continuing may result in data corruption");
+		}
+		else if (perc < 10.0)
+		{
+			m_logger->error("Available disk space is becoming low, please consider releasing more disk space");
+		}
+		if (m_lastPerc2 > 0.0)
+		{
+			double diff = m_lastPerc2 - perc;
+			if (diff > 0.0)
+			{
+				double prediction = (perc * CHECK_THRESHOLD)/ (3600.0 * diff);
+				if (prediction <= 72.0 && (m_lastPrediction2 == 0.0 || m_lastPrediction2 - prediction > 1.0))
+				{
+					m_logger->error("At current rates disk space will be exhausted in %.0f hours", prediction);
```

Comment:
"rate"?

Reply:
It would either be "At the current rate" or "At current rates". Since it could be two storage plugins consuming disk space, and maybe other files such as buckets, using rates rather than rate is probably more correct.